### PR TITLE
Ty 1932 part 1 improve dev tool training

### PR DESCRIPTION
**Pull request by rustonaut**
Based on PR #162. 

I'm splitting up the follow up PR before it was even created.

This contains some refactoring as well as support for running training in parallel.

### CODE REVIEW

File: dev-tool/src/list_net/convert.rs [hunk 1]
```diff
@@ -51,27 +54,100 @@ impl ConvertCmd {
...
         } = self;
 
-        let mut storage = InMemorySamples::default();
-
-        for entry in fs::read_dir(&from_soundgarden)? {
-            let path = entry?.path();
-            if path.extension() != Some(OsStr::new("csv")) {
-                continue;
-            }
+        // Estimate Number of samples based on the number of users.
```
Comment (Robert-Steiner) [inline dev-tool/src/list_net/convert.rs:57]:
```suggestion
        // Estimate the number of samples based on the number of users.
```

File: dev-tool/src/list_net/convert.rs [hunk 2]
```diff
@@ -51,27 +54,100 @@ impl ConvertCmd {
...
+}
 
-        debug!("Write output file.");
-        storage.serialize_into_file(to_samples)?;
+fn list_csv_files(
+    dir: impl AsRef<Path>,
+) -> Result<impl Iterator<Item = Result<PathBuf, Error>>, Error> {
+    let err_msg = "Reading dir with user data-frames failed.";
+    let iter = fs::read_dir(&dir)
+        .context(err_msg)?
```
Comment (Robert-Steiner) [inline dev-tool/src/list_net/convert.rs:128]: Can't we just have string here and remove `err_msg`?

File: dev-tool/src/list_net/data_source.rs [hunk 3]
```diff
@@ -94,10 +104,11 @@ where
     NoEvaluationSamples(f32),
     /// A batch size of 0 is not usable for training.
     BatchSize0,
-    /// The batch size is larger then the number of samples.
-    TooLargeBatchSize(usize),
-    /// Not reset/initialized.
-    ResetWasNotCalledBeforeTraining,
+    /// The batch size ({batch_size}) is larger then the number of samples ({nr_training_samples}).
```
Comment (Robert-Steiner) [inline dev-tool/src/list_net/data_source.rs:107]:
```suggestion
    /// The batch size ({batch_size}) is larger than the number of samples ({nr_training_samples}).
```

File: dev-tool/src/list_net/data_source.rs [hunk 4]
```diff
@@ -319,18 +335,50 @@ impl InMemorySamples {
         })
     }
 
-    /// Add a new sample.
+    /// Prepares samples to be added to this storage.
+    ///
+    /// This is split out from adding samples so that the preparation
```
Comment (Robert-Steiner) [inline dev-tool/src/list_net/data_source.rs:340]: What do you mean by `split out`?

File: xayn-ai/src/ltr/list_net/mod.rs [hunk 5]
```diff
@@ -575,29 +575,24 @@ where
     ///
     /// This will use the `list_net`, `data_source`, `callbacks` and `optimizer` used
     /// to create this `ListNetTrainer`.
-    pub fn train(
-        mut self,
-        epochs: usize,
-        batch_size: usize,
-    ) -> Result<C::Outcome, TrainingError<D::Error, C::Error>> {
+    pub fn train(mut self, epochs: usize) -> Result<C::Outcome, TrainingError<D::Error, C::Error>> {
```
Comment (Robert-Steiner) [inline xayn-ai/src/ltr/list_net/mod.rs:578]: I think we can remove the batch size part in the doc string of this method

File: xayn-ai/src/ltr/list_net/mod.rs [hunk 6]
```diff
@@ -664,15 +652,18 @@ where
...
         }
 
-        let count = costs.len() as f32;
-        let mean =
-            (count > 0.).then(|| costs.into_iter().fold(0f32, |acc, cost| acc + cost / count));
-        Ok(mean)
+        self.callbacks
+            .end_of_evaluation()
+            .map_err(TrainingError::Control)?;
+        Ok(())
```
Comment (Robert-Steiner) [inline xayn-ai/src/ltr/list_net/mod.rs:666]:
```suggestion
            .map_err(TrainingError::Control)
```

File: xayn-ai/src/ltr/list_net/mod.rs [hunk 7]
```diff
@@ -730,11 +719,20 @@ pub trait TrainingController {
...
+        batch: Vec<Sample>,
+        map_fn: impl Fn(Sample) -> (GradientSet, f32) + Send + Sync,
+    ) -> Vec<GradientSet>;
+
     /// Called before started training a batch.
     fn begin_of_batch(&mut self) -> Result<(), Self::Error>;
 
     /// Called after training a batch, the loss for each sample will be passed in.
-    fn end_of_batch(&mut self, losses: Vec<f32>) -> Result<(), Self::Error>;
+    fn end_of_batch(&mut self) -> Result<(), Self::Error>;
```
Comment (Robert-Steiner) [inline xayn-ai/src/ltr/list_net/mod.rs:735]: the doc string needs to be updated

File: dev-tool/src/list_net/convert.rs [hunk 8]
```diff
@@ -51,27 +54,100 @@ impl ConvertCmd {
...
 
-        let mut storage = InMemorySamples::default();
-
-        for entry in fs::read_dir(&from_soundgarden)? {
-            let path = entry?.path();
-            if path.extension() != Some(OsStr::new("csv")) {
-                continue;
-            }
+        // Estimate Number of samples based on the number of users.
+        let nr_users = progress_spin_until_done("Estimating number of samples", || {
```
Comment (Robert-Steiner) [inline dev-tool/src/list_net/convert.rs:58]:
```suggestion
        let nr_users = progress_spin_until_done("Estimating the number of samples", || {
```

File: xayn-ai/src/ltr/list_net/mod.rs [hunk 9]
```diff
@@ -743,11 +741,13 @@ pub trait TrainingController {
     ///
     /// The passed in reference to `list_net` can be used to e.g. bump the intermediate training
     /// result every 10 epochs.
-    fn end_of_epoch(
-        &mut self,
-        list_net: &ListNet,
-        mean_kl_divergence_evaluation: Option<f32>,
-    ) -> Result<(), Self::Error>;
+    fn end_of_epoch(&mut self, list_net: &ListNet) -> Result<(), Self::Error>;
```
Comment (Robert-Steiner) [inline xayn-ai/src/ltr/list_net/mod.rs:744]: the doc string needs to be updated

File: xayn-ai/src/ltr/list_net/mod.rs [hunk 10]
```diff
@@ -743,11 +741,13 @@ pub trait TrainingController {
...
+
+    fn begin_of_evaluation(&mut self, nr_samples: usize) -> Result<(), Self::Error>;
```
Comment (Robert-Steiner) [inline xayn-ai/src/ltr/list_net/mod.rs:746]: Maybe we can also add a bit of documentation for the new methods

**Review by Robert-Steiner — APPROVED** [review]
Looks good to me 👍
Only added a few minor improvements/questions

File: xayn-ai/src/ltr/list_net/mod.rs (hunk 10 above)
Comment (Robert-Steiner) [reply xayn-ai/src/ltr/list_net/mod.rs:746]: just saw that you remove those methods in #164

File: dev-tool/src/list_net/data_source.rs (hunk 4 above)
Comment (rustonaut) [reply dev-tool/src/list_net/data_source.rs:340]: I will update the documentation, originally this was done "on-the-fly" when adding samples, but for supporting parallelization it's easier to do it in a separate step and then add it.  (Also allows of testing both things independently.)

File: dev-tool/src/list_net/convert.rs (hunk 2 above)
Comment (rustonaut) [reply dev-tool/src/list_net/convert.rs:128]: Given how the code is yes,

but the code should have (and did at some point) use the error message for
the case `entry?` or `file_type()` fail...

File: dev-tool/src/list_net/cli_callbacks.rs [hunk 11]
```diff
@@ -73,20 +101,40 @@ impl CliTrainingController {
...
 
+    fn run_batch(
+        &mut self,
+        batch: Vec<Sample>,
+        map_fn: impl Fn(Sample) -> (GradientSet, f32) + Send + Sync,
+    ) -> Vec<GradientSet> {
+        let losses = Mutex::new(Vec::new());
+        let gradient_sets = batch
+            .into_iter()
+            .par_bridge()
```
[the remark this answers is not in the document]
Comment (rustonaut) [reply dev-tool/src/list_net/cli_callbacks.rs:114]: By now none, but there are some types (e.g. arbitrary std  `Iterator` instances) which don't implement `IntoParalleIterator` and at some point batch was such a type.

File: xayn-ai/src/ltr/list_net/mod.rs (hunk 6 above)
Comment (rustonaut) [reply xayn-ai/src/ltr/list_net/mod.rs:666]: Can I keep it the way it is?

We semantically return no error, and changing it to return the `Ok` value of `end_of_evaluation()` does to me convey "we return whatever `.end_of_evaluation()`" returns (which just happens to be `()`).

**Comment by rustonaut** [comment]
There are two thinks for which a mini re-review would be nice:

1. The place where I rewrote the documentation.
2. The one small question I asked later on.

**Comment by Robert-Steiner** [comment]
> There are two thinks for which a mini re-review would be nice:
> 
>     1. The place where I rewrote the documentation.
> 
>     2. The one small question I asked later on.

Have you pushed the changes because the last commit is 4 days old

File: xayn-ai/src/ltr/list_net/mod.rs (hunk 6 above)
Comment (Robert-Steiner) [reply xayn-ai/src/ltr/list_net/mod.rs]: I'm fine keeping it that way 👍

**Review by Robert-Steiner — APPROVED** [review]
LGTM 🚀

**Pull request merged**